tools/codegen/dest/ufunc.py

```diff
@@ -0,0 +1,440 @@
+            for dtype in loops[k].supported_dtypes:
+                ufunctor_sigs.setdefault(dtype, {})[k] = ufunctor_sig
+            continue
+
+        # Otherwise, look in ANY of the generic entries.  For simplicity of
+        # codegen, both ScalarOnly and Generic are defined, the ufunc name
+        # must match  (if they didn't match, we'd have to generate distinct
+        # functors per dtype, which is awful, so we're not going to do it unless
+        # someone really forces us to)
+        ufunc_name = None
+        supported_dtypes = set()
+        for lk in [UfuncKey.ScalarOnly, UfuncKey.Generic]:
+            if ufunc_name is None:
+                ufunc_name = loops[lk].name
+            else:
+                assert ufunc_name == loops[lk].name, "ScalarOnly and Generic must have same ufunc name"
+            supported_dtypes |= loops[lk].supported_dtypes
+        assert ufunc_name is not None
+
+        name = f"{k}_{ufunc_name}"
+        ufunctor_sig = UfunctorSignature(g, scalar_tensor=scalar_tensor_lookup[k], name=name)
+        for dtype in supported_dtypes:
+            ufunctor_sigs.setdefault(dtype, {})[k] = ufunctor_sig
+
+        ufunc_sig = UfuncSignature(g, name=f"ufunc::{ufunc_name}", compute_t=BaseCType(opmath_t))
+        apply_ctx = ufunctor_sig.fields() + ufunctor_sig.arguments().apply
+        ufunctors.append(f"""
+template <typename scalar_t>
+struct {ufunctor_sig.name} {{
+  using opmath_t = at::opmath_type<scalar_t>;
+  {ufunctor_sig.decl_fields()}
+  {ufunctor_sig.inline_defn_ctor()}
+  __device__ {ufunctor_sig.decl_apply()} {{
+    return {ufunc_sig.call(apply_ctx)};
+  }}
+}};
+""")
+
+    return ufunctor_sigs, "\n".join(ufunctors)
+
+@dataclass(frozen=True)
+class BinaryScalarSpecializationConfig:
+    scalar_idx: int
+    ctor_tensor: str
+    ufunc_key: UfuncKey
+
+BinaryScalarSpecializationConfigs = [
+    BinaryScalarSpecializationConfig(
+        scalar_idx = 0,
+        ctor_tensor = 'self',
+        ufunc_key = UfuncKey.CUDAFunctorOnOther,
+    ),
+    BinaryScalarSpecializationConfig(
+        scalar_idx = 1,
+        ctor_tensor = 'other',
+        ufunc_key = UfuncKey.CUDAFunctorOnSelf,
+    ),
+]
+
+def compute_ufunc_cuda_dtype_body(g: NativeFunctionsGroup, dtype: ScalarType, inner_loops: Dict[UfuncKey, UfunctorSignature], parent_ctx: Sequence[Binding]) -> str:
+    body = "using opmath_t = at::opmath_type<scalar_t>;"
+    body += "if (false) {}\n";
+    for config in BinaryScalarSpecializationConfigs:
+        if config.ufunc_key not in inner_loops:
+            continue
+        ufunctor_sig = inner_loops[config.ufunc_key]
+        scalar_idx = config.scalar_idx + 1
+        # Make a copy and at the same time widen the type (not permissible
+        # without copy; we don't want to mutate the input argument anyway)
+        ctx: List[Union[Expr, Binding]] = [b for b in parent_ctx]
+        ctx.append(Expr(
+            expr=f"iter.scalar_value<opmath_t>({scalar_idx})",
```

Comment:
This comment also more just me thinking out loud than providing real feedback: in the code output from the description, the constructor takes in both `iter.scalar_value<opmath_t>(1)` and `(alpha).to<opmath_t>()) `. That second expression gets figured out automatically by `translate()`, but the first needs to be manually added to the context.

I guess we could avoid that split and make more things automatic by teaching codegen about the `TensorIteratorBase`, so `translate()` can automatically figure out that expression... but that's probably a heavy lift, and seems like it only provides minor benefits.

Reply:
Yeah, adding this expression to the context here is basically "teaching" translate about this expression but not teaching it anything about TensorIterator. This is literally the only code generated iter access, so let's wait until there's more :)